Feature/#57 permission schema + Feature/#28 Activity Stream

cmfcmf opened the pull request:
This does a few things. It is recommended to review the pull request commit by commit.
- Install the DataDog AuditBundle to log changes in a better way (for each change, log 'old' and 'new' value, correctly log manytomany associations).
- Add a simple backend to view the activity stream - it is mostly copied from the bundle's examples folder.
- Pass the real Adventure Entity, not the ElasticSearch Adventure Document to the `show` page to be able to display monster types.
- Implement the permission system discussed in #57. This includes the removal of the 'Be a Curator' button
- Correctly configure the EasyAdmin backend to only allow actions which don't break the site (i.e. deleting users)
- Restrict access to user table to admins
- Set access permission for TagContent / TagName controllers to admin, to make sure nobody tinkers with them. Probably can remove the controllers and tables entirely once the data scheme has been converted
- Add getter and setter for createdBy/updatedBy fields to be able to display them in the EasyAdmin backend
- Remove `NotBlank` assertion from plainPassword field, Symfony checks this automatically during registration and it's impossible to update users (i.e. change roles) otherwise. **EDIT:** I lied, the check is still needed. I moved it into the form, though (9980cc3). Therefore the `plainPassword` field is only checked during registration and not when the user is updated through the admin interface.

#### Further things
- Set everyone's roles to `ROLE_USER`, only Matt and Jerry are `ROLE_ADMIN`.
- Fix bugs for the Audit Log filtering
- Add environments to show page
- Clarify BBEG are also monsters
- Highlight currently selected menu item

harkenn commented on app/config/config.yml:
```diff
@@ -111,18 +111,135 @@ stof_doctrine_extensions:
             class: AppBundle\Entity\User
             label: 'Users'
+            form:
+                help: 'Please make sure the user always either has ROLE_USER, ROLE_USER&ROLE_CURATOR or ROLE_USER&ROLE_CURATOR&ROLE_ADMIN!'
```
I'm pretty sure this isn't actually right? Symfony docs indicate that when using role_hierarchy, that roles are inherited implicitly. Our security.yml:
```
    role_hierarchy:
        ROLE_CURATOR: ROLE_USER
        ROLE_ADMIN:  [ROLE_CURATOR, ROLE_ALLOWED_TO_SWITCH]
```
[Symfony docs on the subject](http://symfony.com/doc/current/security.html#hierarchical-roles):
> In the above configuration, users with ROLE_ADMIN role will also have the ROLE_USER role. The ROLE_SUPER_ADMIN role has ROLE_ADMIN, ROLE_ALLOWED_TO_SWITCH and ROLE_USER (inherited from ROLE_ADMIN).

I think its actually supposed to be only **one** of `ROLE_USER`, `ROLE_CURATOR` or `ROLE_ADMIN` without the need for compositing multiple roles together.

harkenn commented on src/AppBundle/Security/AdventureVoter.php:
```diff
@@ -0,0 +1,117 @@
+     */
+    private function canEditAndDelete(Adventure $adventure, TokenInterface $token, User $user)
+    {
+        if ($this->decisionManager->decide($token, ['ROLE_CURATOR'])) {
```
Could be simplified to
```
return (
    $this->decisionManager->decide($token, ['ROLE_CURATOR']) 
    || $adventure->getCreatedBy() === $user->getUsername()
)
```

cmfcmf replied:
You're right, of course. I opted for multiple checks to make this highly important piece of code as readable and easy to understand as possible.

harkenn commented on src/AppBundle/Security/AdventureVoter.php:
```diff
@@ -0,0 +1,117 @@
+    protected function voteOnAttribute($attribute, $subject, TokenInterface $token)
+    {
+        // Everyone can view all adventures
+        if ($attribute === self::VIEW) {
```
This isn't ever going to change, I don't think? I'd probably advocate stripping out `AdventureVoter::VIEW` across the board.

cmfcmf replied:
If we implement drafts, for example, not everyone would be able to see all adventures. I added it for completeness. To me the `AdventureVoter` feels right if it is capable of voting on all possible actions for an adventure.

cmfcmf commented on src/AppBundle/Controller/AuditController.php:
```diff
@@ -0,0 +1,103 @@
+                break;
+            default:
+                // if user attempts to filter by other fields, we restrict it
+                throw new \Exception("filter not allowed");
```
You're talking about the query builder variable, right? I can add a comment to make it more explicit. Maybe I shouldn't have copy-pasted all that much code from the DataDog examples ;)

harkenn replied:
Yes, the `$qb` variable

cmfcmf commented:
Don't merge yet, I'm still working on your comments :)

cmfcmf commented:
Addressed the comments, also adjusted a few more minor things I noticed when migrating the data:
- Set everyone's roles to `ROLE_USER`, only Matt and Jerry are `ROLE_ADMIN`.
- Fix bugs for the Audit Log filtering
- Add environments to show page
- Clarify BBEG are also monsters
- Highlight currently selected menu item

harkenn commented:
Actually looks like you had since taken the offending string out. So looks good!
